src/contracts/core/StakeRootCompendium.sol

```diff
@@ -221,137 +253,152 @@ contract StakeRootCompendium is StakeRootCompendiumStorage {
-        );
+        uint256 _chargePerProof =
+            operatorSets.length * chargeParams.chargePerOperatorSet + totalStrategies * chargeParams.chargePerStrategy;
+        require(_chargePerProof <= chargeParams.maxChargePerProof, ChargePerProofExceedsMax());
+        chargePerProof.push(uint32(block.timestamp), uint224(_chargePerProof));
     }
 
-    function _cumulativeCharges() internal view returns (uint32, uint96, uint96) {
+    function _calculateCumulativeCharges() internal view returns (uint32, uint96, uint96) {
         // calculate the total charge since the last update up until the latest calculation timestamp
         // note that there may be no corresponding stakeRootSubmission for the latest calculation timestamp
         // but if the calculationTimestamp is in the past, then it should be charged for, since proofs are being generated
-        uint32 latestCalculationTimestamp = uint32(block.timestamp) - uint32(block.timestamp % proofIntervalSeconds);
-        if (cumulativeChargeLastUpdatedTimestamp == latestCalculationTimestamp) {
-            return (latestCalculationTimestamp, cumulativeChargePerOperatorSetLastUpdate, cumulativeChargePerStrategyLastUpdate);
+        uint32 latestCalculationTimestamp =
+            uint32(block.timestamp) - uint32(block.timestamp % cumulativeChargeParams.proofIntervalSeconds);
+
+        if (cumulativeChargeParams.lastUpdateTimestamp == latestCalculationTimestamp) {
+            return (
+                latestCalculationTimestamp, cumulativeChargeParams.chargePerOperatorSet, cumulativeChargeParams.chargePerStrategy
+            );
         }
-        uint256 numProofs = (latestCalculationTimestamp - cumulativeChargeLastUpdatedTimestamp) / proofIntervalSeconds;
+
+        uint256 numProofs = (latestCalculationTimestamp - cumulativeChargeParams.lastUpdateTimestamp)
+            / cumulativeChargeParams.proofIntervalSeconds;
+
         return (
-            latestCalculationTimestamp, 
-            uint96(cumulativeChargePerOperatorSetLastUpdate + chargePerOperatorSet * numProofs), 
-            uint96(cumulativeChargePerStrategyLastUpdate + chargePerStrategy * numProofs)
+            latestCalculationTimestamp,
+            uint96(cumulativeChargeParams.chargePerOperatorSet + chargeParams.chargePerOperatorSet * numProofs),
+            uint96(cumulativeChargeParams.chargePerStrategy + chargeParams.chargePerStrategy * numProofs)
         );
     }
 
     function _updateCumulativeCharge() internal {
-        (uint32 latestCalculationTimestamp, uint96 cumulativeChargePerOperatorSet, uint96 cumulativeChargePerStrategy) = _cumulativeCharges();
-        cumulativeChargeLastUpdatedTimestamp = latestCalculationTimestamp;
-        cumulativeChargePerOperatorSetLastUpdate = cumulativeChargePerOperatorSet;
-        cumulativeChargePerStrategyLastUpdate = cumulativeChargePerStrategy;
+        (uint32 lastUpdateTimestamp, uint96 cumulativeChargePerOperatorSet, uint96 cumulativeChargePerStrategy) =
```

Comment:
maybe check `lastUpdateTimestamp` against `cumulativeChargeParams.lastUpdateTimestamp` and return early?